mantidproject/mantid
# Rotate CSGObjects and coordinate with SetGoniometer and CopySample

The following shapes can be rotated, and the finite ones can be plotted:
Cuboid
Cylinder
HollowCylinder
InfinitePlane
InfiniteCylinder
Hexahedron
TaperedGuide
SliceofCylinderRing

The rotation can be manually set by the user using the rotate tag
Further rotation is added using a goniometer tag, which works with SetGoniometer

CopySample can now copy a shape which can be plotted, but maintains its own
Goniometer angle

**Description of work.**

**To test:**

1) Run the below script. To rotate the cuboid around the x,y,z axes (in that order) change the values in the rotate tag in `cuboid_xml`. Check this works for [other shapes too](https://docs.mantidproject.org/nightly/concepts/HowToDefineGeometricShape.html#howtodefinegeometricshape), and that infinite cylinders and infinite cuboids cannot be plotted!

<details><summary>Rotate CSGObjects Script</summary>
<p>

``` python
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np
from mpl_toolkits.mplot3d.art3d import Poly3DCollection
from mantid.api import AnalysisDataService as ADS

ws = CreateSampleWorkspace()

cuboid_xml = " \
<cuboid id='some-cuboid'> \
  <height val='20.0'  /> \
  <width val='50.0' />  \
  <depth  val='20.0' />  \
  <centre x='100.0' y='100.0' z='100.0'  />  \
  <rotate x='0' y='0' z='0' /> \
</cuboid>  \
<algebra val='some-cuboid' /> \
"

# cylinder_xml = ' \
# <cylinder id="stick"> \
#   <centre-of-bottom-base x="-1.0" y="0.0" z="0.0" /> \
#   <axis x="1.0" y="0.0" z="0.0" /> \
#   <radius val="0.2" /> \
#   <height val="2.0" /> \
#   <rotate x=\"0\" y=\"0\" z=\"0\" /> \
# </cylinder> \
# <algebra val=\'some-cylinder\' /> \
# '
# # 
# merge_xml = ' \
# <cylinder id="stick"> \
#   <centre-of-bottom-base x="-0.5" y="0.0" z="0.0" /> \
#   <axis x="1.0" y="0.0" z="0.0" />  \
#   <radius val="0.05" /> \
#   <height val="1.0" /> \
# </cylinder> \
#  \
# <sphere id="some-sphere"> \
#   <centre x="0.0"  y="0.0" z="0.0" /> \
#   <radius val="0.5" /> \
# </sphere> \
#  \
# <algebra val="some-sphere (# stick)" /> \
# '
# # 
# slice = ' \
# <slice-of-cylinder-ring id="A"> \
#   <inner-radius val="0.0" /> \
#   <outer-radius val="0.0646" /> \
#   <depth val="0.01" /> \
#   <arc val="90.0" />  \
#   <rotate x="90.0" y="45.0" z="180.0" /> \
# </slice-of-cylinder-ring> \
# '
# inf_cylinder = ' \
# <infinite-cylinder id="A" > \
#   <centre x="0.0" y="0.2" z="0" /> \
#   <axis x="0.0" y="0.2" z="0" /> \
#   <radius val="1" /> \
# </infinite-cylinder> \
# '
# tapered-guide = ' \
# <tapered-guide id="A Guide"> \
#     <aperture-start height="2.0" width="2.0" /> \
#     <length val="3.0" /> \
#     <aperture-end height="4.0" width="4.0" /> \
#     <centre x="0.0" y="5.0" z="10.0" /> \  \
#     <axis x="0.0" y="0.0" z="1.0" /> \  \
# </tapered-guide> \
# '

# Set sample geometry of workspace to this CSG object
SetSample(ws, Geometry={'Shape': 'CSG', 'Value': cuboid_xml})

sample = ws.sample()
shape = sample.getShape()
mesh = shape.getMesh()

print(mesh)

facecolors = ['b','royalblue','red','firebrick', 'orange', 'gold','green', 'darkgreen','grey','black', 'purple', 'mediumorchid']

mesh_polygon_a = Poly3DCollection(mesh, facecolors = facecolors, linewidths=0.1)
'''Problem based on order of being added!'''

fig, axes = plt.subplots(subplot_kw={'projection':'mantid3d'})
axes.add_collection3d(mesh_polygon_a)

# Auto scale to the mesh size
axes_lims = mesh.flatten()
axes.auto_scale_xyz(axes_lims, axes_lims, axes_lims)

axes.set_title('Sample Shape: Tube')
axes.set_xlabel('X / m')
axes.set_ylabel('Y / m')
axes.set_zlabel('Z / m')

plt.show()
```
</p>
</details>

Fixes #31449

2) Here is a script to test that CopySample now works with these new rotations by the goniometer. The first setGoniometer on ws does no rotation, but SetGoniometer on ws1 rotates about the y-axis clockwise by 90 degrees. CopySample should be able to copy over the shape (cuboid), but correctly use the goniometer angles for ws1.

<details><summary>SetGoniometer and CopySample for CSGObjects Script</summary>
<p>

``` python
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np
from mpl_toolkits.mplot3d.art3d import Poly3DCollection
from mantid.api import AnalysisDataService as ADS


# A fake host workspace, replace this with your real one.
ws = CreateSampleWorkspace()
SetGoniometer(ws, Axis0="0,0,1,0,1")

# Specify a Sphere geometry using CSG

cuboid_alt_xml = " \
<cuboid id='some-cuboid'> \
  <height val='20.0'  /> \
  <width val='50.0' />  \
  <depth  val='20.0' />  \
  <centre x='100.0' y='100.0' z='100.0'  />  \
  <rotate x='0' y='0' z='0' /> \
</cuboid>  \
<algebra val='some-cuboid' /> \
"

# Set sample geometry of workspace to this CSG object Sphere
SetSample(ws, Geometry={'Shape': 'CSG', 'Value': cuboid_alt_xml})

ws1 = CreateSampleWorkspace()
SetGoniometer(ws1, Axis0="90,0,1,0,1")
CopySample(ws,ws1,CopyEnvironment=False, CopyMaterial=False,CopyShape=True)
 
for i,val in enumerate((ws,ws1)): 
    sample = val.sample()
    shape = sample.getShape()
    mesh = shape.getMesh()

    facecolors = ['royalblue','b','red','firebrick', 'gold', 'orange','green', 'darkgreen','grey','black', 'purple', 'mediumorchid']

    mesh_polygon = Poly3DCollection(mesh, facecolors = facecolors, linewidths=0.1)

    fig, axes = plt.subplots(subplot_kw={'projection':'mantid3d'})
    axes.add_collection3d(mesh_polygon)

    # Auto scale to the mesh size
    axes_lims = mesh.flatten()
    axes.auto_scale_xyz(axes_lims, axes_lims, axes_lims)

    axes.set_title('Sample Shape: Cuboid ws{}'.format(i))
    axes.set_xlabel('X / m')
    axes.set_ylabel('Y / m')
    axes.set_zlabel('Z / m')

    plt.show()
```
</p>
</details>

3) Here is a script to check that the CopySample implementation also works for MeshObjects now too. You will find tube.stl on the systemtest data.

<details><summary>CopySample for MeshObjects Script</summary>
<p>

``` python
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

from mpl_toolkits.mplot3d.art3d import Poly3DCollection
from mantid.api import AnalysisDataService as ADS

# load sample shape mesh file for a workspace
ws = CreateSampleWorkspace()
# alternatively: ws = Load('filepath') or ws = ADS.retrieve('ws')
SetGoniometer(ws, Axis0="0,0,1,0,1")
ws = LoadSampleShape(ws, "tube.stl")

ws1 = CreateSampleWorkspace()
SetGoniometer(ws1, Axis0="90,0,1,0,1")
CopySample(ws,ws1,CopyEnvironment=False, CopyMaterial=False,CopyShape=True)

for i,val in enumerate((ws,ws1)): 
    sample = val.sample()
    shape = sample.getShape()
    mesh = shape.getMesh()

    facecolors = ['royalblue','b','red','firebrick', 'gold', 'orange','green', 'darkgreen','grey','black', 'purple', 'mediumorchid']

    mesh_polygon = Poly3DCollection(mesh, facecolors = facecolors, linewidths=0.1)

    fig, axes = plt.subplots(subplot_kw={'projection':'mantid3d'})
    axes.add_collection3d(mesh_polygon)

    # Auto scale to the mesh size
    axes_lims = mesh.flatten()
    axes.auto_scale_xyz(axes_lims, axes_lims, axes_lims)

    axes.set_title('Sample Shape: Cuboid ws{}'.format(i))
    axes.set_xlabel('X / m')
    axes.set_ylabel('Y / m')
    axes.set_zlabel('Z / m')

    plt.show()
```
</p>
</details>
